cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.metadata.lineage;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.AbstractDataset;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.common.app.RunIds;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.proto.Id;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.collect.ImmutableSet;
+import org.apache.twill.api.RunId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Dataset to store/retrieve Dataset accesses of a Program.
+ */
+// TODO: set conflict-level to none on this dataset.
+public class LineageDataset extends AbstractDataset {
+  private static final Logger LOG = LoggerFactory.getLogger(LineageDataset.class);
+
+  private static final byte[] STOP_COL_BYTE = {'s'};
+  private static final byte[] METADATA_COLS_BYTE = {'m'};
+
+  private static final char DATASET_MARKER = 'd';
+  private static final char PROGRAM_MARKER = 'p';
+  private static final char FLOWLET_MARKER = 'f';
+  private static final char NONE_MARKER = '0';
+
+  Table accessRegistryTable;
+
+  public LineageDataset(String instanceName, Table accessRegistryTable) {
+    super(instanceName, accessRegistryTable);
+    this.accessRegistryTable = accessRegistryTable;
+  }
+
+  public void addAccess(Id.Run run, Id.DatasetInstance datasetInstance,
+                        AccessType accessType, long stopTime, @Nullable Id component) {
```

Comment:
Instead of requiring caller to pass in `null`, it's better to have a overloaded method that doesn't take `component`.